src/runtime/HalideBuffer.h

```diff
@@ -617,7 +617,11 @@ class Buffer {
      * the buffer. Does nothing if this buffer did not allocate its
      * own memory. */
     void deallocate() {
-        decref();
+        if (manages_memory()) {
```

Comment:
I think it's decref that's broken. It claims to null out alloc, but only does that if the count hits zero. decref_dev should similarly null out dev_ref_count unconditionally